swiftnav/src/time/gnss.rs

```diff
@@ -0,0 +1,816 @@
+        seconds: f64,
+    ) -> GpsTime {
+        MJD::from_date(year, month, day, hour, minute, seconds).to_gps_hardcoded()
+    }
+    /// Converts a Rust GPS time to its C equivalent
+    pub(crate) fn to_gps_time_t(self) -> swiftnav_sys::gps_time_t {
+        swiftnav_sys::gps_time_t {
+            tow: self.tow,
+            wn: self.wn,
+        }
+    }
+
+    /// Gets the week number
+    pub fn wn(&self) -> i16 {
+        self.wn
+    }
+
+    /// Gets the time of week
+    pub fn tow(&self) -> f64 {
+        self.tow
+    }
+
+    /// Checks if the stored time is valid
+    pub fn is_valid(&self) -> bool {
+        self.tow.is_finite()
+            && self.tow >= 0.0
+            && self.tow < consts::WEEK_SECS as f64
+            && self.wn >= 0
+    }
+
+    /// Normalize time of week value so it's within the length of a week
+    fn normalize(&mut self) {
+        while self.tow < 0.0 {
+            self.tow += consts::WEEK_SECS as f64;
+            self.wn -= 1;
+        }
+
+        while self.tow >= consts::WEEK_SECS as f64 {
+            self.tow -= consts::WEEK_SECS as f64;
+            self.wn += 1;
+        }
+    }
+
+    /// Adds a duration to the time
+    pub fn add_duration(&mut self, duration: &Duration) {
+        self.tow += duration.as_secs_f64();
+        self.normalize();
+    }
+
+    /// Subtracts a duration from the time
+    pub fn subtract_duration(&mut self, duration: &Duration) {
+        self.tow -= duration.as_secs_f64();
+        self.normalize();
+    }
+
+    /// Gets the difference between this and another time value in seconds
+    pub fn diff(&self, other: &Self) -> f64 {
+        let dt = self.tow - other.tow;
+        dt + (self.wn - other.wn) as f64 * consts::WEEK_SECS as f64
+    }
+
+    /// Convert a [`GpsTime`] into a [`UtcTime`] using the [`UtcParams`] if
+    /// available, or hardcoded leap seconds otherwise
+    fn internal_to_utc(self, params: Option<&UtcParams>) -> UtcTime {
+        // Is it during a (positive) leap second event
+        // Get the UTC offset at the time we're converting
+        let (is_lse, dt_utc) = params.map_or_else(
+            || {
+                (
+                    self.is_leap_second_event_hardcoded(),
+                    self.gps_utc_offset_hardcoded(),
+                )
+            },
+            |p| (self.is_leap_second_event(p), self.gps_utc_offset(p)),
+        );
+
+        let mut tow_utc = self.tow - dt_utc;
+
+        if is_lse {
+            /* positive leap second event ongoing, so we are at 23:59:60.xxxx
+             * subtract one second from time for now to make the conversion
+             * into yyyy/mm/dd HH:MM:SS.sssss format, and add it back later */
+            tow_utc -= 1.0;
+        }
+
+        let mut t_u = GpsTime {
+            wn: self.wn,
+            tow: tow_utc,
+        };
+        t_u.normalize();
+
+        /* break the time into components */
+        let mut u = UtcTime::from_gps_no_leap(t_u);
```

Comment:
This is a huge nit but i find single letter variables really confusing. Can we call this `utc_time` and the `t_u` variable above `gps_time` (or something a bit more descriptive)?

Reply:
Valid point. Changed.